### PR TITLE
Time based tenure extend

### PR DESCRIPTION
**Pull request by obycode**
See #5476

### CODE REVIEW

File: clarity/src/vm/costs/mod.rs
Comment (jbencin) [inline clarity/src/vm/costs/mod.rs:1336]: I would recommend doing:

```suggestion
        *self == Self::zero()
```

It's shorter, and this way if any fields are added to `ExecutionCost`, there's no chance of forgetting to update this function and it returning an incorrect value

Also declare the `zero()` function as `const`, which might help the compiler optimize this code better

File: stacks-signer/src/v0/signer.rs
Comment (jbencin) [inline stacks-signer/src/v0/signer.rs:998]:
```suggestion
        let Ok(Some(mut block_info)) = self.signer_db.block_lookup(block_hash).inspect_err(|e| {
            warn!("{self}: Failed to load block {block_hash}: {e:?})");
        }) else {
```

File: clarity/src/vm/costs/mod.rs
Comment (obycode) [reply clarity/src/vm/costs/mod.rs:1336]: Good point. I guess a `ZERO` const would be even better than a `const` method, right?

File: stackslib/src/net/api/postblock_proposal.rs
Comment (jbencin) [inline stackslib/src/net/api/postblock_proposal.rs:582]: This should be rewritten to use `std::time::Instant`:

```rust
let start = Instant::now();
// ...
let validation_time_ms = start.elapsed();
```

`Instant` uses `CLOCK_MONOTONIC` rather than `CLOCK_REALTIME`. The difference is that `CLOCK_REALTIME` tracks wall clock time, so it can jump around or go backwards due to NTP or other adjustments, whereas `CLOCK_MONOTONIC` cannot go backwards. This makes `CLOCK_MONOTONIC` is better for measuring time elapsed

I know I wrote this function initially, so sorry about that!

File: clarity/src/vm/costs/mod.rs
Comment (obycode) [reply clarity/src/vm/costs/mod.rs]: Refactored to use `ZERO` in 2e266a20622cb4722d0743be3cb180a8d5e2569e.

File: stackslib/src/net/api/postblock_proposal.rs
Comment (obycode) [reply stackslib/src/net/api/postblock_proposal.rs]: Updated in 6e06712f23a1b0bd695c929065385a5991409fc3.

File: clarity/src/vm/costs/mod.rs
Comment (jbencin) [inline clarity/src/vm/costs/mod.rs:1197]: You could have done 

```rust
pub const fn zero() { ... }
```

I think using a constant is better practice, but changing it does make the diff larger. No need to change anything, just commenting

File: clarity/src/vm/costs/mod.rs
Comment (jbencin) [reply clarity/src/vm/costs/mod.rs]: I think it's better practice to use a constant, but this does make the diff bigger. Your call, either way is fine

File: testnet/stacks-node/src/nakamoto_node/stackerdb_listener.rs
Comment (jbencin) [inline testnet/stacks-node/src/nakamoto_node/stackerdb_listener.rs:293]: Minor style issue:

```suggestion
                        let Some(block) = blocks.get_mut(&block_sighash) else {
                            info!(
                                "StackerDBListener: Received signature for block that we did not request. Ignoring.";
                                "signature" => %signature,
                                "block_signer_sighash" => %block_sighash,
                                "slot_id" => slot_id,
                                "signer_set" => self.signer_set,
                            );
                            continue;
                        };
```

File: testnet/stacks-node/src/nakamoto_node/stackerdb_listener.rs
Comment (jbencin) [inline testnet/stacks-node/src/nakamoto_node/stackerdb_listener.rs:246]: I would guess the compiler is smart enough to make this optimization, but this could be:

```suggestion
                .into_iter()
```

File: testnet/stacks-node/src/nakamoto_node/stackerdb_listener.rs
Comment (jbencin) [inline testnet/stacks-node/src/nakamoto_node/stackerdb_listener.rs:48]: `lazy_static!()` is kinda considered deprecated now, and we have an open PR to remove it completely (#5055). Use `std::sync::LazyLock` instead:

```suggestion
/// Fault injection flag to prevent the miner from seeing enough signer signatures.
/// Used to test that the signers will broadcast a block if it gets enough signatures
pub static TEST_IGNORE_SIGNERS: LazyLock<TestFlag> = LazyLock::new(TestFlag::default);
```

File: testnet/stacks-node/src/tests/signer/v0.rs
Comment (jbencin) [inline testnet/stacks-node/src/tests/signer/v0.rs:2831]: Doesn't matter as much since it's in tests, but still should be replaced with `std::time::Instant`

Comment (obycode) [reply testnet/stacks-node/src/tests/signer/v0.rs:2831]: Does it make sense to use `Instant` when we're just trying to get the difference in time between a u64 passed in as a number of seconds since the epoch and now? This one seems correct as-is.

Comment (obycode) [reply testnet/stacks-node/src/tests/signer/v0.rs:2831]: c195c68e54ada6eb784671c963fc0275f46a3e53

File: testnet/stacks-node/src/nakamoto_node/stackerdb_listener.rs
Comment (obycode) [reply testnet/stacks-node/src/nakamoto_node/stackerdb_listener.rs]: c195c68e54ada6eb784671c963fc0275f46a3e53

File: testnet/stacks-node/src/tests/signer/v0.rs
Comment (jbencin) [reply testnet/stacks-node/src/tests/signer/v0.rs:2831]: I could be wrong here. I don't know if there is anywhere else that we can reduce wall clock usage and replace with `Instant`

**Review by jbencin — APPROVED** [review]

File: testnet/stacks-node/src/tests/signer/v0.rs
Comment (obycode) [reply testnet/stacks-node/src/tests/signer/v0.rs:2831]: I think in this case, because the time we're comparing against is a wall clock time that we received, we have no choice but to use wall clock time.

File: stacks-signer/CHANGELOG.md
Comment (jferrant) [inline stacks-signer/CHANGELOG.md:15]: Such a nit: but should prob say "when a time-based tenure extend will be accepted" since there are other reasons we might issue one.

**Review by jcnelson — APPROVED** [review]
Just a minor comment about the signer DB, but that can be addressed in a separate PR.

**Review by jbencin — DISMISSED** [review]

File: stacks-signer/CHANGELOG.md
Comment (obycode) [reply stacks-signer/CHANGELOG.md]: 👍 3bccb233c

File: stacks-signer/src/signerdb.rs
Comment (obycode) [reply stacks-signer/src/signerdb.rs:945]: No. Agreed, I think we should remove `block_info` in a separate PR.

**Review by jbencin — DISMISSED** [review]

**Comment by jcnelson** [comment]
One last thing I think we ought to do here -- can you set the block version to something that isn't 0 or 1?  Maybe 2?  Then, we can observe the rollout.

**Review by jbencin — DISMISSED** [review]

**Review by jcnelson — DISMISSED** [review]

**Review by hstove — APPROVED** [review]

**Comment by obycode** [comment]
Failing tests all pass locally, except for `tests::nakamoto_integrations::mock_mining`. I'm checking on that now.

**Comment by obycode** [comment]
I pushed two changes related to mock-mining:
1. Do not propose block when mock-mining -- this was a new problem, caused by these changes
2. Do not broadcast the block when mock-mining -- this was an existing problem

The mock-mining test is back to being flaky now.

File: testnet/stacks-node/src/nakamoto_node/miner.rs
Comment (jferrant) [inline testnet/stacks-node/src/nakamoto_node/miner.rs:694]: This is a nit and was already in place with the mining_key check before so don't hold up the PR for it, but maybe we should move these config checks to the very top of the function. Will this potentially prevent unnecessary deadlocks for the chainstate and sortition db?

Comment (obycode) [reply testnet/stacks-node/src/nakamoto_node/miner.rs:694]: good point. May as well move them up.

Comment (obycode) [reply testnet/stacks-node/src/nakamoto_node/miner.rs:685]: Moved in 5bedd6e43.

File: testnet/stacks-node/src/nakamoto_node/miner.rs
Comment (jferrant) [inline testnet/stacks-node/src/nakamoto_node/miner.rs:430]: another nit so again don't hold up for it: could we get this indented block into some sort of function call ? Function is kinda long and its hard to read with the added indentation. Maybe just even put the match section into a function call like like handle_block_proposal_result or something.

**Review by jferrant — APPROVED** [review]

File: testnet/stacks-node/src/nakamoto_node/miner.rs
Comment (obycode) [reply testnet/stacks-node/src/nakamoto_node/miner.rs]: b201dcda2

**Review by jferrant — APPROVED** [review]

**Review by hstove — APPROVED** [review]

**Review by jcnelson — APPROVED** [review]

**Review by jbencin — APPROVED** [review]

**Pull request merged**
